internal/cephfs/controllerserver.go

```diff
@@ -1119,3 +1119,267 @@ func deleteSnapshotAndUndoReservation(
 
 	return nil
 }
+
+// ControllerPublishVolume implements the CSI ControllerPublishVolume RPC.
+func (cs *ControllerServer) ControllerPublishVolume(
+	ctx context.Context,
+	req *csi.ControllerPublishVolumeRequest,
+) (*csi.ControllerPublishVolumeResponse, error) {
+	// If the CSI driver is not running on Kubernetes, we do not need to
+	// unfence the node, so we return immediately.
+	// This is because the unfencing logic is specific to Kubernetes environments.
+	if !k8s.RunsOnKubernetes() {
+		return &csi.ControllerPublishVolumeResponse{}, nil
+	}
+
+	if req.GetVolumeId() == "" {
+		return nil, status.Error(codes.InvalidArgument, "Volume ID cannot be empty")
+	}
+	if req.GetNodeId() == "" {
+		return nil, status.Error(codes.InvalidArgument, "Node ID cannot be empty")
+	}
+	if req.GetVolumeCapability() == nil {
+		return nil, status.Error(codes.InvalidArgument, "Volume Capabilities cannot be empty")
+	}
+	secrets := req.GetSecrets()
+
+	if acquired := cs.VolumeLocks.TryAcquire(req.GetVolumeId()); !acquired {
+		return nil, status.Errorf(codes.Aborted, util.VolumeOperationAlreadyExistsFmt, req.GetVolumeId())
+	}
+	defer cs.VolumeLocks.Release(req.GetVolumeId())
+
+	err := cs.unfenceNode(ctx, req.GetNodeId(), req.GetVolumeId(), secrets, cs.Driver.IsFencingEnabled())
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return &csi.ControllerPublishVolumeResponse{}, nil
+}
+
+// ControllerUnpublishVolume implements the CSI ControllerUnpublishVolume RPC.
+func (cs *ControllerServer) ControllerUnpublishVolume(
+	ctx context.Context,
+	req *csi.ControllerUnpublishVolumeRequest,
+) (*csi.ControllerUnpublishVolumeResponse, error) {
+	// If the CSI driver is not running on Kubernetes, we do not need to
+	// fence the node, so we return immediately.
+	// This is because the fencing logic is specific to Kubernetes environments.
+	if !k8s.RunsOnKubernetes() {
+		return &csi.ControllerUnpublishVolumeResponse{}, nil
+	}
+
+	if req.GetVolumeId() == "" {
+		return nil, status.Error(codes.InvalidArgument, "Volume ID cannot be empty")
+	}
+	// NodeId is optional in the ControllerUnpublishVolumeRequest.
+	if req.GetNodeId() == "" {
+		return nil, status.Error(codes.InvalidArgument, "Node ID cannot be empty")
+	}
```

Comment:
Am not sure this is correct check as it can have side effects in case CO doesnt set it as this field is optional. How about we dont do fencing if this is empty and return success from here as we cannot do fencing?( This is to keep today behavior and to be on safer side)